proof.go

```diff
@@ -154,20 +161,20 @@ func (proof Proof) VerifyNamespace(h hash.Hash, nID namespace.ID, data [][]byte,
 		// collect leaf hashes from provided data and do some sanity checks:
 		hashLeafFunc := nth.HashLeaf
 		for _, gotLeaf := range data {
-			if len(gotLeaf) < int(nIDLen) {
-				// conflicting namespace sizes
+			if nth.ValidateLeaf(gotLeaf) != nil {
 				return false
 			}
-			gotLeafNid := namespace.ID(gotLeaf[:nIDLen])
-			if !gotLeafNid.Equal(nID) {
+			// check whether the namespace ID of the data matches the queried nID
+			if gotLeafNid := namespace.ID(gotLeaf[:nIDLen]); !gotLeafNid.Equal(nID) {
 				// conflicting namespace IDs in data
 				return false
 			}
-			leafData := append(
-				gotLeafNid, gotLeaf[nIDLen:]...,
-			)
 			// hash the leaf data
-			gotLeafHashes = append(gotLeafHashes, hashLeafFunc(leafData))
+			leafHash, err := hashLeafFunc(gotLeaf)
+			if err != nil { // this can never happen due to the initial validation of the leaf at the beginning of the loop
+				return false
```

Comment:
[optional] if this can never happen, should it `panic`?

Reply:
The reason it cannot happen is that the leaf is validated earlier in the code. However, if that section of the code changes in the future, we may encounter an error at this line. If the leaf is found to be invalid, we should return false instead of panicking.

Reply:
Even when you check the line where the leaf gets validated, you will see that we return false.